### ISSUE TITLE
[PVC] Prebuild without PVC, user A account with PVC and trigger rebuild, user B account without PVC

### ISSUE THREAD

**Issue by jenting** [gitpod-io/gitpod#12718]
### Bug description

The prebuild project _without_ PVC.

The user A account _with_ PVC, add new branch X and commit, push branch X to the repo, trigger prebuild on Gitpod. The prebuild runs _without_ PVC.
The user B account _without_ PVC, open that branch X, the user B uses PVC.

### Steps to reproduce

The prebuild projects _without_ PVC.

- The user A account _with_ PVC.
- The user B account _without_ PVC.
- User A adds new branch X and commit, pushes branch X to the repo, and triggers prebuild on Gitpod. The prebuild runs _without_ PVC. 👍 
- User B opens branch X, the user B uses PVC (The prebuild project and the user's account are both _without_ PVC). 

### Expected behavior

User B should open the workspace _without_ PVC because the prebuild project is configured _without_ PVC, also the user's account _does not enable_ PVC.

### Anything else?

#7901

**Comment by kylos101** [comment]
:wave:  @jenting I just tested this, and got same result. Here are supporting details:

project has pvc disabled, https://gitpod.io/t/prebuild-experiment/prebuild-experiment

user has PVC enabled, https://gitpod.io/admin/users/8df3495b-685d-46e0-9820-009cc3b4afd8

prebuild did not use PVC, which is expected, because it's disabled on the project. Details:
```
node: headless-ws-us64-internal-xl-pool-nhjr
workspace: kylos101-prebuildexperi-5p1e6ngpkxf
pod: prebuild-880d87db-2e38-45e1-8ada-5bfa79f94210
```

Same user has PVC disabled now

PVC is still disabled on the project

A new workspace is started from the branch's prebuild, but has PVC. Details:
```
node: workspace-ws-us64-large-pvc-pool-kfwz
workspace: kylos101-prebuildexperi-biberrbv8vk
pod: ws-f3d2657c-4e47-4ee1-aca3-c557e6bcc9e3 
```

**Comment by jenting** [comment]
I think when open the workspace, we should check the prebuild project PVC enabled or the user's PVC enabled.

If one of the condition hits, we send start workspace request with feature flag `persistent_volume_claim`. Otherwise, the start workspace request must not contains feature flag `persistent_volume_claim`.

**Comment by kylos101** [comment]
:wave: @jenting ,

> I think when we open the workspace, we should check the prebuild project PVC enabled or the user's PVC enabled.

A couple questions:
1. Do you mean prebuild workspace or regular workspace?
2. Why check prebuild project PVC setting, or user PVC feature flag?

Considerations:
1. the PVC setting on the project (we check this now, it controls if the prebuild runs with PVC or not)
2. the PVC feature flag on the user (we check this now, it controls if new regular workspaces start with PVC or not)
3. the feature flag persisted to the prebuild workspace config field in the database (if this is set, we should start regular workspaces with PVC, I'd argue the problem is we're wrongly setting this...especially given the scenario I shared :point_up:, and to fix, maybe we have to avoid persisting it in the first place when the project has PVC disabled )
4. the feature flag persisted to the regular workspace config field in the database (this should control whether a regular workspace restart uses PVC or not :crossed_fingers: )

It seems like the problem is, when we run prebuilds, if the startWorkspace request contains `persistent_volume_claim`, and the project has PVC disabled, we're persisting `persistent_volume_claim` to the workspace's config JSON object in the database. So, as a result, when we start regular workspaces from the prebuild, they use PVC (because the config object says it was used).

**Comment by jenting** [comment]
> 1. Do you mean prebuild workspace or regular workspace?

regular workspace 

> 2. Why check prebuild project PVC setting, or user PVC feature flag?

Because if the project was built with PVC, then for any user who open the regular workspace from that prebuild will force to use PVC.

Because the user enables PVC, open workspace from prebuild (with PVC or without PVC) or from git clone, the regular workspace should use PVC.

**Comment by kylos101** [comment]
I like your idea, @jenting . My only concern with user feature flag in its current form, is that it is hard to roll out. For example, we have to set it via Admin page, or SQL query.

Ideally we'd set the `persistent_volume_claim` feature flag for regular workspace start from the `server` component using rules that we define in configcat. This way its not exclusively at the user level.

Checkout https://github.com/gitpod-io/gitpod/issues/12745 for the idea, should help us rollout in a more flexible way.

**Comment by sagor999** [comment]
Hey @geropl 
We need webapp help on this one.
tl;dr
starting prebuild via gitpod.io/#prebuild/<repo>
incorrectly (in my opinion) bundles prebuild and workspace together, specifically, feature flags from workspace are somehow getting coalesced with prebuild itself.

Here is a correct scenario:
User pushes a commit into repo
User triggers prebuild (or it gets triggered automatically).
Prebuild completes. All is good.

Broken scenario:
User pushes a commit into repo
User immediately triggers prebuild via `#prebuild/` url 
It runs prebuild, and then immediately opens workspace from that prebuild.
Another user opens that prebuild, and feature flags from workspace get embedded into prebuild.

This is where it happens:
https://github.com/gitpod-io/gitpod/blob/40b89ce1eefae3ca821da8ff837b9022d6d60a25/components/server/src/workspace/gitpod-server-impl.ts#L1155

This seems to me like a bug, and we should not tangle prebuild and workspace together in that scenario, as we are leaking feature flags of a user into prebuild. 

Can you provide insight? If it is indeed a bug, would you mind scheduling a fix for it?

**Comment by geropl** [comment]
Hey @sagor999, :wave: 

let me re-phrase, to make sure we're on the same page:
 - you see a prebuild configured by - and running with credentials from - user A
 - that prebuilds is using all features flags/config from user A (<-- incl. PVC flag)
 - when user B gets around and opens a workspace based on that prebuild, it breaks for them, because they cannot use the PVC workspace class

If that's a correct summary this is expected behavior. :confused: I don't say it's ideal, and we need to fix the hard prebuild <-> user coupling (see "domain model" discussion where having soon).

I'm not sure if we support the PVC -> old workspace backup path, but from the context I assume not. Then I see two ways:
 - we limit the PVC workspace flag to `regular` workspace
 - we allow "propagation" of the PVC flag, e.g. it becomes infectious (I guess that's what we don't want to do to not loose control over the rollout)

**Comment by sagor999** [comment]
Yeah, correct summary.

**Comment by kylos101** [comment]
@geropl 

Given this block
https://github.com/gitpod-io/gitpod/blob/5bae14df896a36f92fce7ca423ae22db32c16383/components/server/src/workspace/workspace-starter.ts#L851-L860

It seems like the problem could be that we're passing `pvcEnabledForPrebuilds` as true in the options, when really, it's `false` for the project, but true the user has the `persistent_volume_claim` feature flag set.

I see options are set here:
https://github.com/gitpod-io/gitpod/blob/5bae14df896a36f92fce7ca423ae22db32c16383/components/server/src/workspace/workspace-starter.ts#L302

But the only place where they are being set explicitly is here:
https://github.com/gitpod-io/gitpod/blob/5bae14df896a36f92fce7ca423ae22db32c16383/components/server/ee/src/prebuilds/prebuild-manager.ts#L255

**Comment by sagor999** [comment]
@kylos101 no, this is not correct.
`pvcEnabledForPrebuilds` is set to true ONLY when project has PVC enabled. 

So in our case prebuild is correctly ran without PVC (since project has it disabled), but then it would open workspace with PVC enabled since user has PVC flag enabled, and then that flag would get baked into prebuild somehow. And this only happens when running prebuild via `#prebuild` url.

**Comment by kylos101** [comment]
:+1: ah, okay, I was unsure if I had the right block/path/flow, thank you @sagor999 for clarifying! @sagor999 it might help to sync with @geropl or another on webapp team. Could you share a calendly link with Gero?

**Comment by jankeromnes** [comment]
> And this only happens when running prebuild via `#prebuild` url.

Whoops, the `#prebuild` URL prefix is deprecated and should be removed soon, because it bypasses pretty much everything we've built in the past years: https://github.com/gitpod-io/gitpod/issues/4353

The supported ways to trigger Prebuilds are (by order of preference):
1. Push a new commit to a branch with Prebuilds enabled
2. Click on the "Run Prebuild" button in Gitpod's UI for a failed or cancelled prebuild

**Comment by jenting** [comment]
🤔 but I remember my reproduce steps are
- push a new commit to a branch
- Click the "Run Prebuild" button in Gitpod's UI

**Comment by sagor999** [comment]
@jenting can you try to repro this? I was only able to repro this when using #prebuild url. 🤔

**Comment by geropl** [comment]
I think I found the bug; it is as @jenting described.

The problem:
 - [here](https://github.com/gitpod-io/gitpod/blob/7a23e97cf64fc2fe59e957fcf3a2dcde8a9aa40f/components/server/src/workspace/config-provider.ts#L122-L135), we define `persistent_volume_claim` as "persistent" for a workspace
 - [here](https://github.com/gitpod-io/gitpod/blob/1ba4edc7fcc8f89c8d406ef0bb3843235817ed34/components/server/ee/src/workspace/workspace-factory.ts#L158) it (consequently) copies over from prebuild -> regular workspace (based on that prebuild)
 - ( [here](https://github.com/gitpod-io/gitpod/blob/1eb228e937166f2ce0029c916d5ca31520a5c922/components/server/src/workspace/workspace-starter.ts#L831-L833) we break the idea of "feature flags persistent for the lifetime of a workspace" ... but fixing this might be out-of-scope for in this context )
 - [here](https://github.com/gitpod-io/gitpod/blob/1eb228e937166f2ce0029c916d5ca31520a5c922/components/server/src/workspace/workspace-starter.ts#L849-L860) we don't filter that PVC feature flag from `regular` workspaces

Just to make sure I understand how `user.featureFlags` and `project?.settings?.usePersistentVolumeClaim` are meant to interact:
 - user's control how whether they want to use PVC or not
 - projects control whether prebuilds use PVC or not
 - we support the transition "bucket backup" -> "PVC backup"
 - we do not support the reverse

D'accord? :thinking: 

Based on this I'll create PR.
